### PR TITLE
importer/kv:  add ImportEpoch field to MVCCValueHeader and write to it during IMPORT INTO

### PR DESCRIPTION
**Pull request by msbutler**
storage: add ImportEpoch field to MVCCValueHeader
This patch adds the ImportEpoch field to an MVCCValue's MVCCValueHeader,
which allows kv clients (namely the sst_batcher in an IMPORT INTO) to write
the importing table's ImportEpoch to the metadata of each ingesting MVCCValue.

Unlike the MVCCValueHeader.LocalTimestamp field, the ImportEpoch field should
be exported to other clusters (e.g. via ExportRequests from BACKUP/RESTORE and
streaming). Consequently, this PR relaxes the invariant that the
MVCCValueHeader field must be stripped in an Export Request and must be empty
in an AddSSTable Request. Now, Export Request only strips the
MVCCValueHeader.LocalTimestamp field and AddSSTable will only require the
LocalTimestamp to be empty.

Release note: none

----

bulk/kv write the table's ImportEpoch to each MVCCValue during IMPORT
This patch makes IMPORT INTO on a non-empty table write the table's ImportEpoch
to each ingested MVCC Value, via the SSTBatcher. In a future PR, the
ImportEpoch will be used to track and rollback an IMPORT INTO. This additional
information will allow IMPORTing tables to be backed up and restored, as
described in this [RFC](https://docs.google.com/document/d/16TbkFznqbsu3mialSw6o1sxOEn1pKhhJ9HTxNdw0-WE/edit#heading=h.bpox0bmkz77i).

Informs https://github.com/cockroachdb/cockroach/issues/76722

Release note: None

### CODE REVIEW

**Comment by msbutler** [comment]
This is ready for a look!
 - Curious if y'all think I should break this into a few separate PRs. 
 - All the current test failures relate to incorrect mvcc stats getting returned in data driven tests. This is expected since this patch makes an `MVCCValue` two bytes larger. I'll update these tests once y'all have given the PR a pass.

File: pkg/kv/bulk/sst_batcher_test.go
Comment (msbutler) [inline pkg/kv/bulk/sst_batcher_test.go:358]: the test below covers a subset of the test in import_mvcc_test.go. I include in case we want to divide this PR.

**Comment by msbutler** [comment]
@erikgrinaker Responding to a few points you made:
>might be a good idea to have a utility method MVCCValueHeader.StripForExport()
Done.
>Might make sense with a couple of commits
Done.
On the MVCCValue becoming two bytes larger: Something fishy is going on. If I run `TestEncodeDecodeMVCCValue` with the folllowing modification: 

https://github.com/cockroachdb/cockroach/blob/aaa711b216e4fa48c1882ce07cb5a198990df0a7/pkg/storage/mvcc_value.go#L190-L198

This prints out: 
```
Client Meta Size 0
Local Timestamp Size 2
MVCC Value Header Size (ClientMeta + Local Timestamp):6
```
Does it have something to do with this ?
https://github.com/cockroachdb/cockroach/blob/aaa711b216e4fa48c1882ce07cb5a198990df0a7/pkg/storage/mvcc_value.go#L173-L187

**Comment by erikgrinaker** [comment]
Ah, I was thinking more about this bit here, which just omits the entire header if it doesn't contain anything:

https://github.com/cockroachdb/cockroach/blob/aaa711b216e4fa48c1882ce07cb5a198990df0a7/pkg/storage/mvcc_value.go#L204-L208

So that seems fine.

>On the MVCCValue becoming two bytes larger: Something fishy is going on.

I think you're missing the encoding of the embedded messages. Both `util.hlc.Timestamp` and `ClientMeta` are embedded Protobuf messages. Essentially, these are encoded with the tag (field ID + write type) and length prefix, in addition to the actual contents. For encoding details, see: https://developers.google.com/protocol-buffers/docs/encoding

Let's have a look at the encoding here. The `TestEncodeDecodeMVCCValue` case is:

```go
MVCCValueHeader{
  LocalTimestamp: hlc.ClockTimestamp{WallTime: 9},
}
```

For reference, the Protobuf schemas are:

```protobuf
message MVCCValueHeader {
  util.hlc.Timestamp local_timestamp = 1;
  ClientMeta client_meta = 2;
}

message ClientMeta {
  int64 import_job_id = 1;
}

message Timestamp {
  int64 wall_time = 1;
  int32 logical = 2;
  bool synthetic = 3;
}
```

The encoded value before we added the `ClientMeta` field was this:

```
0x0a 0x02 0x08 0x09
```

* `0x0a`: field 1 of wire type delimited (i.e. the embedded `Timestamp` message) → (1 << 3) + 2 = 10 = 0x0a
* `0x02`: the length of the message in field 1 → 2 bytes
* `0x08`: field 1 of wire type varint in `Timestamp` (i.e. `WallTime`) → (1 << 3) + 0 = 8 = 0x08
* `0x09`: the varint value of `Timestamp.WallTime` → 9

Ok, so far so good. We see that it omitted fields 2 and 3 of `Timestamp` because they were empty, so that's great.

Now let's add an empty `ClientMeta` field. The encoded value is now this:

```
0x0a 0x02 0x08 0x09 0x12 0x00
```

Ok, so the first 4 bytes are the same as before. The 2 added bytes are:

* `0x12`: field 2 of wire type delimited (i.e. the embedded `ClientMeta` message) → (2 << 3) + 2 = 18 = 0x12
* `0x00`: the length of the message in field 2 → 0 bytes

Ok, so that makes sense -- it's saying that there's an embedded `ClientMeta` message that's empty. Now, since empty fields can be omitted (we see this both with the `Timestamp` and `ClientMeta` contents), it's a bit surprising that it's including the `ClientMeta` message field without any contents. That seems entirely useless, and is probably a bug/inefficiency in the gogoproto encoder -- I had a look at the generated code, and it does in fact unconditionally include nested messages. I'll add an issue to look into that separately, because it might give us some minor savings all over the place.

But in any case, this isn't terrible, because it's still omitting the actual message contents. So even if we're paying a fixed cost of 2 bytes for the message, at least we can keep adding on fields to `ClientMeta` without it increasing the encoded size if they're unset. So I think this is probably fine.

FWIW, we can also see this with the `Timestamp`: if it's empty, we're still paying a 2-byte cost for it when encoded.

**Comment by msbutler** [comment]
wow, thanks for really digging into this! I learned a bunch from this write up. My takeaways are:
- Protobuf encodings do _not_ omit empty messages, but do omit empty scale types (and potentially other empty proto types). We're unsure why, but we could address this in our own code.
- Given all of this, we expect my diff to increase the size of an MVCCValueHeader, therefore, all these tests that report MVCC stats need to be updated.

**Review by erikgrinaker — COMMENTED** [review]
Would be good to specify the overhead here, specifically the typical number of bytes added per key, as well as the per-value encoding CPU/latency overhead.

**Review by msbutler — COMMENTED** [review]
I think I addressed everything your latest round of feedback except for reporting benchmark results. Should I report those in the commit message or in the docstring? As expected, the latency percent increase for Encode/DecodeMVCC value varies depending on the amount of data already in the MVCC value. When I benchmarked EncodedMVCCValue for example, a header with a JobId was 22% slower than an empty header with a `short` encoded MVCCValue.value, and 4% slower with a `long` encoded MVCCValue.value.

File: pkg/kv/kvserver/kvserverbase/bulk_adder.go
Comment (erikgrinaker) [reply pkg/kv/kvserver/kvserverbase/bulk_adder.go:76]: Hm, an interesting wrinkle here is that we can restore into a different cluster, and `GenerateUniqueInt` isn't guaranteed to be unique across clusters, since it's based on the node ID and timestamp. That actually makes collisions far more likely than if they were completely random, since the values are very clustered.

This is mostly academic though, since it would need a pretty unlikely sequence of events.

Also, these IDs are going to be fairly large since they populate the higher-order bits of the ID (which means varints won't help). So we're adding on something like 10 bytes to each key.

File: pkg/kv/bulk/sst_batcher.go
Comment (erikgrinaker) [reply pkg/kv/bulk/sst_batcher.go:252]: Makes sense, thanks.

File: pkg/storage/mvcc.go
Comment (erikgrinaker) [reply pkg/storage/mvcc.go]: Nice! Let's call it 'EncodeForExport()` or something then.

File: pkg/ccl/backupccl/restore_data_processor.go
Comment (erikgrinaker) [reply pkg/ccl/backupccl/restore_data_processor.go:486]: Hmmm, interesting. I wonder if we can get away with not writing the import IDs in the general case her. This is speculative, but bear with me.

The only reason we need these import IDs is to handle restore at current timestamp. During backup, we know which imports are in flight. We can also know their start time. Could we add on the import IDs during export? In fact, if there can only be 1 import in flight per table, is it enough with a boolean (latter may have issues with chained backup/restores)?

Comment (msbutler) [reply pkg/ccl/backupccl/restore_data_processor.go:486]: aha! I really like the idea of _only_ writing the ImportID during Export Request (or any other kv request used during streaming). While you're right that only 1 import can be in flight per table, we can't merely write a bool to the metadata because of incremental backups. Consider the following situation:

t0: Import A begins on foo
t1: Full backup of foo begins
t2: Import A ends 
t3: Import B begins on foo
t4: incremental backup of foo begins
t5: restore, only import B is still in progress

Suppose during backup, we only wrote an `IsImporting` bool to each key in the backup SSTs: the full backup would write that bool to some keys from import A, and the incremental backup would write the bool to some keys from import B. During restore, however, we'd have no way to distinguish between keys written during Import A vs B. So, each backed up, imported key, needs a unique ID. An easy unique ID is the job ID, but if we wanted to, we could potentially get away with writing a unique `short` to each key during Export. 

I.e.
- During import bind the shortID and the JobId to each table descriptor
- During backup planning:  gather all the unique shortIDs from active imports
- During export: write the shortID to each importing key
- During restore: create a map[oldShortID]NewShortID`, rewrite the restored descriptor and import job with the newShortID.

Comment (erikgrinaker) [reply pkg/ccl/backupccl/restore_data_processor.go:486]: Yep, makes sense. I don't think it's necessarily worth shortening the IDs, but if we could get away with only writing them for the rare cases where we're backing up in-flight imports, the savings would probably be worth it in the typical case.

Wdyt @dt?

Comment (msbutler) [reply pkg/ccl/backupccl/restore_data_processor.go:486]: hmm. During an unrelated meeting, David pointed out the sad fact that our current approach of _only_ writing the ImportID during an Export Request falls apart for Backups/Restore's _of_ tenants, conducted by the system tenant. In this case, the system tenant has no access to the tenant's sql descriptors, and consequently, its importing job Ids. Therefore, there's currently no way to plumb the tenant's importing job IDs to the export request. I wonder though, the system tenant is so powerful! Is there an easy way to gather tenant descriptors during backup planning?

That being said, I think the original plan I proposed still works for tenant backup/restores, since the jobs table is completely preserved. This implies that there's no need to track and rewrite importing jobIDs for tenants, as there's no chance of jobID collision between the backed up and restored tenant.

**Comment by erikgrinaker** [comment]
> Should I report those in the commit message or in the docstring?

When making changes like these, it's usually nice to explain the cost in the commit message. What the reader/reviewer really wants to know is how expensive this is going to be. How much additional data are we storing? Are imports now 1% slower? 10% slower? Not really slower at all? Numbers like these help paint a picture of the cost.

22% slower encoding might matter, but how much does the encoding cost make up of the per-key processing cost? The absolute value per encoding might be helpful to understand the which order of magnitude we're talking (i.e. nanoseconds vs milliseconds), although I'm guessing we're talking O(100 ns) per key here.

File: pkg/ccl/backupccl/restore_data_processor.go
Comment (erikgrinaker) [reply pkg/ccl/backupccl/restore_data_processor.go:486]: Aw, that's a bummer. Ok, back to the original plan then, I suppose.

File: pkg/kv/kvserver/kvserverbase/bulk_adder.go
Comment (msbutler) [reply pkg/kv/kvserver/kvserverbase/bulk_adder.go:76]: I'm not sure I follow. Per this [comment](https://github.com/cockroachdb/cockroach/pull/85138#discussion_r934733926), I think the restoring cluster should create new unique job IDs for each in progress import. I'll verify this plan in my comment with david.

Comment (erikgrinaker) [reply pkg/kv/kvserver/kvserverbase/bulk_adder.go:76]: The "unique" ID is simply a concatenation of the node ID and current time, so it's only unique within a single cluster (because we guarantee clock monotonicity via the HLC). It's therefore possible (but unlikely) for e.g. two restores from two different imports in two different clusters to have the same ID, or for the target cluster to be running with a slow clock and generate the same ID.

I think this is mostly academic though, for a variety of reasons -- I mostly found it interesting/amusing that a random UUID would be far less likely to collide, because it isn't clustered around the current time and very-low-integer node IDs.

We don't need to do anything here.

File: pkg/ccl/backupccl/restore_data_processor.go
Comment (dt) [reply pkg/ccl/backupccl/restore_data_processor.go:486]: Well, we could just add some range state:
before you start importing, you send an rpc to the span you will import into saying the time above which you'll be importing and your import ID. distsender fans this out to every range in that span, and each of them persists this mapping in the range state. During splits it is duplicated to both sides, etc just like any other range state.
When we export eval, we can read the range state and apply the ID to any key with ts above the assigned value. Or we could just return the mapping from export at that point, instead of doing anything key-by-key.

I mean, this sorta sounds like range keys?

Comment (dt) [reply pkg/ccl/backupccl/restore_data_processor.go:486]: also sounds a lot like our lock-span-for-op plans

Comment (erikgrinaker) [reply pkg/ccl/backupccl/restore_data_processor.go:486]: Hmm, that's an interesting idea. It's definitely doable, although we've currently hardcoded an assumption that MVCC range keys are range tombstones throughout the entire system since we didn't know what live range keys would look like yet. Probably a bit late in the game to change that with a week left before stability (and a ton of other work to do before then), even though I don't think it should be a huge lift to do something like that later.

Comment (msbutler) [reply pkg/ccl/backupccl/restore_data_processor.go:486]: Then, when the import job is synthesized during RESTORE, this range state operation would happen again with the new job ID.

File: pkg/storage/mvcc.go
Comment (msbutler) [reply pkg/storage/mvcc.go]: Done.

**Comment by msbutler** [comment]
Given yesterday's meeting, here's the updated design:
- In the `MVCCValueHeader`, scrap the `ClientMeta` struct and write an `ImportEpoch` to each MVCCValueHeader, which identifies the number of times a user has called IMPORT INTO on the table this key belongs to when the table was not empty. This is more space efficient than the original design outlined in the RFC.
- This metadata will get written during an IMPORT INTO on a non empty table.
- In a separate PR that should merge before this, I'll add the ImportEpoch to the sql descriptor.

File: pkg/ccl/backupccl/restore_data_processor.go
Comment (msbutler) [reply pkg/ccl/backupccl/restore_data_processor.go]: GIven meeting yesterday, here's the [new approach](https://github.com/cockroachdb/cockroach/pull/85138#issuecomment-1205780978).

File: pkg/kv/bulk/sst_batcher.go
Comment (erikgrinaker) [inline pkg/kv/bulk/sst_batcher.go:266]: We may want to consider adding a function to decode-modify-encode an MVCCValue (taking a callback), so that we can make use of the faster inlined functions. But we can do that after this PR too.

File: pkg/storage/enginepb/mvcc3.proto
Comment (erikgrinaker) [inline pkg/storage/enginepb/mvcc3.proto:175]: nit: consider briefly explaining its purpose, i.e. that it's used for cancelling an import following a restore.

File: pkg/storage/mvcc_value.go
Comment (erikgrinaker) [inline pkg/storage/mvcc_value.go:154]: nit: since we moved this to a top-level function rather than a method, let's do `EncodeMVCCValueForExport()`.

**Review by erikgrinaker — APPROVED** [review]
This all looks good, once the table descriptor change is in.

**Comment by msbutler** [comment]
closing, since for 22.2, we decided that we can rollback imports with mvcc timestamps.

**Comment by stevendanna** [comment]
Reopened and rebased. Let's see what CI says.

**Comment by stevendanna** [comment]
@erikgrinaker You approved an old version of this, but you may want to take another look.

**Comment by stevendanna** [comment]
NB: I am going to add this new option to the MVCCPut, et al functions in the PR that adds a new delete range predicate. But I'm happy to make one big PR if that is easier for y'all.

**Review by dt — APPROVED** [review]

**Comment by dt** [comment]
> one big PR if that is easier

eh, I say merge this one as-is now and add follow-ups to follow-ups. I easier to review that way and we’ll start getting some nightly runs.

File: pkg/storage/mvcc_value_test.go
Comment (msbutler) [inline pkg/storage/mvcc_value_test.go:248]: could you add the bench results to the commit msg?

File: pkg/storage/mvcc_value.go
Comment (msbutler) [inline pkg/storage/mvcc_value.go:240]: nit: could you add a unit test for this?

**Review by msbutler — COMMENTED** [review]
Looks great! could you run this through the mixed version roachtest before merging?

File: pkg/storage/mvcc_value.go
Comment (stevendanna) [reply pkg/storage/mvcc_value.go:240]: Done. Piggybacked on the tests for DecodeMVCCValue so that we are always testing the same cases for both.

Comment (msbutler) [reply pkg/storage/mvcc_value.go:240]: Thanks!

**Review by msbutler — COMMENTED** [review]
LGTM! I can't approve because of github shenanigans.

**Comment by erikgrinaker** [comment]
>@erikgrinaker You approved an old version of this, but you may want to take another look.

Sorry, still making my way through the queue, will get to this in the morning.

**Review by erikgrinaker — APPROVED** [review]
LGTM once the minor comments below are addressed. I've only superficially reviewed this, particularly the DR parts.

> Reviewed 4 of 19 files at r6, 7 of 8 files at r7, 2 of 11 files at r8, 15 of 15 files at r17, 10 of 10 files at r18, 3 of 3 files at r19, 7 of 7 files at r20, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/85138#-:-NsmQB-mA_LiKGgbigHn:b9xzazd)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @michae2 and @msbutler)

___
*[`pkg/ccl/backupccl/backup_processor.go` line 485 at r20](https://reviewable.io/reviews/cockroachdb/cockroach/85138#-NsmNt0n6TeLJSi_Bhla:-NsmNt0n6TeLJSi_Bhlb:b-3v1m9d) ([raw file](https://github.com/cockroachdb/cockroach/blob/1ec8a18c2741e6845cd77c86bd529356ca739959/pkg/ccl/backupccl/backup_processor.go#L485)):*
> ```Go
> 				for _, span := range spans {
> 					for len(span.span.Key) != 0 {
> 						includeMVCCValueHeader := clusterSettings.Version.IsActive(ctx, clusterversion.V24_1)
> ```

Is it ok to flip this in the middle of a backup? Can we get into weird states where part of the backup has value headers but the other part doesn't? Can we race with an import starting to write import epochs, but the backup processor not seeing them for part of the exports?

Actually, why do we even need to version gate this? A 23.2 node will ignore the field, but also won't ever export any non-empty value headers. A 24.1 node may have non-empty value headers after the 24.1 gate has switched and an import has started writing them, but at that point all binaries are prepared to handle them anyway. So I think we can just set this unconditionally, and avoid any races altogether?
___
*[`pkg/ccl/backupccl/restore_data_processor.go` line 591 at r19](https://reviewable.io/reviews/cockroachdb/cockroach/85138#-NsmNH682dK9EKQre4Qv:-NsmNH682dK9EKQre4Qw:bembsht) ([raw file](https://github.com/cockroachdb/cockroach/blob/c6c300cd9d772eb221df95471def620f00fb839d/pkg/ccl/backupccl/restore_data_processor.go#L591)):*
> ```Go
> 		// InitChecksum don't copy/reallocate the slice they
> 		// were given.
> 		if err := batcher.AddMVCCKey(ctx, key, valueScratch); err != nil {
> ```

It's not obvious at all here that the checksum rewriting above modifies `valueScratch`, it's worth spelling this out.
___
*[`pkg/kv/kvserver/kvserverbase/bulk_adder.go` line 77 at r18](https://reviewable.io/reviews/cockroachdb/cockroach/85138#-NsmKYQY8UG1iizF4A7l:-NsmKYQZ0_MQYH2Ceg-B:b-z3ohqp) ([raw file](https://github.com/cockroachdb/cockroach/blob/6ed2789c28fc04b620aaf19391864e89ddb3f0d0/pkg/kv/kvserver/kvserverbase/bulk_adder.go#L77)):*
> ```Go
> 	// Adder's SSTBatcher will write the import epoch to each versioned value's
> 	// metadata.
> 	ImportEpoch uint32
> ```

Is it worth calling out here that this must only be used with 24.1 clusters, since it's the caller's responsibility to enforce the version gate?
___
*[`pkg/storage/mvcc_value.go` line 144 at r17](https://reviewable.io/reviews/cockroachdb/cockroach/85138#-NsmIUyZ5a2OmvasUgAQ:-NsmIUyZ5a2OmvasUgAR:b-t5gwdk) ([raw file](https://github.com/cockroachdb/cockroach/blob/9b6f65f40b5e9202be3f5bf396aac7b73d09db4c/pkg/storage/mvcc_value.go#L144)):*
> ```Go
> 	// Consider a fast path, where only the roachpb.Value gets exported.
> 	// Currently, this only occurs if the value was not imported.
> 	if mvccValue.ImportEpoch == 0 {
> ```

This seems a bit brittle, in that we could accidentally omit fields that are added later. Should we invert the fast path such that we first strip fields (i.e. the `LocalTimestamp`) and then compare the stripped value header with `MVCCValueHeader{}` to omit the encode?

File: pkg/storage/mvcc_value.go
Comment (dt) [inline pkg/storage/mvcc_value.go:142]: It is weird to me that we get the whole header or none of it just depending on this one field?

If we're going to condition on this field, it sorta feels like this should be the only field you get, or, more generally, you should only get fields that are in the condition?

File: pkg/sql/importer/import_processor.go
Comment (dt) [inline pkg/sql/importer/import_processor.go:395]: The len()==1 makes me nervous.

What about in the loop body:
```
if bulkAdderImportEpoch == 0 { bulkAdderImportEpoch = v.Desc.ImportEpoch } else if bulkAdderImportEpoch != v.Desc.InportEpoc { return errors.Assertion....
```

**Review by dt — APPROVED** [review]

**Review by stevendanna — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/85138)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @dt, @erikgrinaker, @michae2, and @msbutler)

___
*[`pkg/ccl/backupccl/backup_processor.go` line 485 at r20](https://reviewable.io/reviews/cockroachdb/cockroach/85138#-NsmNt0n6TeLJSi_Bhla:-NsmbwemAIU1P4XrxxLF:b9ed3ql) ([raw file](https://github.com/cockroachdb/cockroach/blob/1ec8a18c2741e6845cd77c86bd529356ca739959/pkg/ccl/backupccl/backup_processor.go#L485)):*
<details><summary><i>Previously, erikgrinaker (Erik Grinaker) wrote…</i></summary><blockquote>

Is it ok to flip this in the middle of a backup? Can we get into weird states where part of the backup has value headers but the other part doesn't? Can we race with an import starting to write import epochs, but the backup processor not seeing them for part of the exports?

Actually, why do we even need to version gate this? A 23.2 node will ignore the field, but also won't ever export any non-empty value headers. A 24.1 node may have non-empty value headers after the 24.1 gate has switched and an import has started writing them, but at that point all binaries are prepared to handle them anyway. So I think we can just set this unconditionally, and avoid any races altogether?
</blockquote></details>

I think we do need the version check here but you are right that we don't want it changing through the life of the backport. I'll add some state to our spec.

We need the version because if our version is 23.2 we expect to be able to restore the artifact on 23.2 nodes, but 23.2 nodes won't be able to read our SSTs.

**Review by stevendanna — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/85138)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @dt, @erikgrinaker, @michae2, and @msbutler)

___
*[`pkg/ccl/backupccl/restore_data_processor.go` line 591 at r19](https://reviewable.io/reviews/cockroachdb/cockroach/85138#-NsmNH682dK9EKQre4Qv:-Nsn1X7X7-c9UzNUoaAY:b-f8sf1p) ([raw file](https://github.com/cockroachdb/cockroach/blob/c6c300cd9d772eb221df95471def620f00fb839d/pkg/ccl/backupccl/restore_data_processor.go#L591)):*
<details><summary><i>Previously, erikgrinaker (Erik Grinaker) wrote…</i></summary><blockquote>

It's not obvious at all here that the checksum rewriting above modifies `valueScratch`, it's worth spelling this out.
</blockquote></details>

Added a comment.
___
*[`pkg/storage/mvcc_value.go` line 144 at r17](https://reviewable.io/reviews/cockroachdb/cockroach/85138#-NsmIUyZ5a2OmvasUgAQ:-Nsn2B7I92SB2Vxi0aGe:b-e9717d) ([raw file](https://github.com/cockroachdb/cockroach/blob/9b6f65f40b5e9202be3f5bf396aac7b73d09db4c/pkg/storage/mvcc_value.go#L144)):*
<details><summary><i>Previously, erikgrinaker (Erik Grinaker) wrote…</i></summary><blockquote>

This seems a bit brittle, in that we could accidentally omit fields that are added later. Should we invert the fast path such that we first strip fields (i.e. the `LocalTimestamp`) and then compare the stripped value header with `MVCCValueHeader{}` to omit the encode?
</blockquote></details>

I keep going back and forth on this. I've changed the code so that you only get this single field exported. But I do see the potential value in exporting everything.

**Review by erikgrinaker — COMMENTED** [review]
> Reviewed 8 of 9 files at r21, 19 of 19 files at r23, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/85138)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @dt, @michae2, @miraradeva, and @msbutler)

___
*[`pkg/ccl/backupccl/backup_processor.go` line 485 at r20](https://reviewable.io/reviews/cockroachdb/cockroach/85138#-NsmNt0n6TeLJSi_Bhla:-Nsn618B2UFirK514Ld-:b7dq212) ([raw file](https://github.com/cockroachdb/cockroach/blob/1ec8a18c2741e6845cd77c86bd529356ca739959/pkg/ccl/backupccl/backup_processor.go#L485)):*
<details><summary><i>Previously, stevendanna (Steven Danna) wrote…</i></summary><blockquote>

I think we do need the version check here but you are right that we don't want it changing through the life of the backport. I'll add some state to our spec.

We need the version because if our version is 23.2 we expect to be able to restore the artifact on 23.2 nodes, but 23.2 nodes won't be able to read our SSTs.
</blockquote></details>

Right, makes sense.
___
*[`pkg/storage/mvcc_value.go` line 144 at r17](https://reviewable.io/reviews/cockroachdb/cockroach/85138#-NsmIUyZ5a2OmvasUgAQ:-Nsn4ps0A-Jd2gF2xjAQ:b-m76chg) ([raw file](https://github.com/cockroachdb/cockroach/blob/9b6f65f40b5e9202be3f5bf396aac7b73d09db4c/pkg/storage/mvcc_value.go#L144)):*
<details><summary><i>Previously, stevendanna (Steven Danna) wrote…</i></summary><blockquote>

I keep going back and forth on this. I've changed the code so that you only get this single field exported. But I do see the potential value in exporting everything.
</blockquote></details>

Ok. What about `omit_in_rangefeeds`, should those not be preserved across a backup/restore? I haven't been involved with that work, so I don't know what semantics we want.

cc @miraradeva

**Comment by stevendanna** [comment]
@erikgrinaker 

> Ok. What about omit_in_rangefeeds, should those not be preserved across a backup/restore? I haven't been involved with that work, so I don't know what semantics we want.

David and I discussed this is chat and for now I think we'll exclude it.  We can perhaps revisit in a follow-up later this week.

**Comment by stevendanna** [comment]
bors r=dt,erikgrinaker

**Comment by stevendanna** [comment]
bors r=dt,erikgrinaker

**Pull request merged**
